File: core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/statistics/TimeRangeStatistic.java
/**
 * Copyright (c) 2013-2022 Contributors to the Eclipse Foundation
 *
 * <p> See the NOTICE file distributed with this work for additional information regarding copyright
 * ownership. All rights reserved. This program and the accompanying materials are made available
 * under the terms of the Apache License, Version 2.0 which accompanies this distribution and is
 * available at http://www.apache.org/licenses/LICENSE-2.0.txt
 */
package org.locationtech.geowave.core.geotime.store.statistics;

import java.util.Calendar;
import java.util.Date;
import org.locationtech.geowave.core.geotime.util.TimeUtils;
import org.locationtech.geowave.core.store.api.DataTypeAdapter;
import org.locationtech.geowave.core.store.api.FieldStatistic;
import org.locationtech.geowave.core.store.api.Statistic;
import org.locationtech.geowave.core.store.entities.GeoWaveRow;
import org.locationtech.geowave.core.store.statistics.field.FieldStatisticType;
import org.threeten.extra.Interval;

public class TimeRangeStatistic extends FieldStatistic<TimeRangeStatistic.TimeRangeValue> {
  public static final FieldStatisticType<TimeRangeValue> STATS_TYPE =
      new FieldStatisticType<>("TIME_RANGE");

  public TimeRangeStatistic() {
    super(STATS_TYPE);
  }

  public TimeRangeStatistic(final String typeName, final String fieldName) {
    super(STATS_TYPE, typeName, fieldName);
  }

  @Override
  public String getDescription() {
    return "Maintains the time range of a temporal field.";
  }

  @Override
  public TimeRangeValue createEmpty() {
    return new TimeRangeValue(this);
  }

  @Override
  public boolean isCompatibleWith(final Class<?> fieldClass) {
    return Date.class.isAssignableFrom(fieldClass)
        || Calendar.class.isAssignableFrom(fieldClass)
        || Number.class.isAssignableFrom(fieldClass);
  }

  public static class TimeRangeValue extends AbstractTimeRangeValue {

    public TimeRangeValue() {
      this(null);
    }

    public TimeRangeValue(final Statistic<?> statistic) {
      super(statistic);
    }

    @Override
    protected <T> Interval getInterval(
        final DataTypeAdapter<T> adapter,
        final T entry,
        final GeoWaveRow... rows) {
      final Object fieldValue =
          adapter.getFieldValue(entry, ((TimeRangeStatistic) statistic).getFieldName());
      return TimeUtils.getInterval(fieldValue);
    }

  }
}

File: python/src/main/python/pygw/statistics/field/time_range_statistic.py
#
# Copyright (c) 2013-2022 Contributors to the Eclipse Foundation

#
# See the NOTICE file distributed with this work for additional information regarding copyright
# ownership. All rights reserved. This program and the accompanying materials are made available
# under the terms of the Apache License, Version 2.0 which accompanies this distribution and is
# available at http://www.apache.org/licenses/LICENSE-2.0.txt
# ===============================================================================================
from pygw.config import geowave_pkg
from ..statistic import FieldStatistic
from ..statistic_type import FieldStatisticType
from ...base.interval import IntervalTransformer


class TimeRangeStatistic(FieldStatistic):
    """
    Tracks the time range of a temporal field.
    """
    STATS_TYPE = FieldStatisticType(geowave_pkg.core.geotime.store.statistics.TimeRangeStatistic.STATS_TYPE)

    def __init__(self, type_name=None, field_name=None, java_ref=None):
        if java_ref is None:
            if type_name is None and field_name is None:
                java_ref = geowave_pkg.core.geotime.store.statistics.TimeRangeStatistic()
            else:
                java_ref = geowave_pkg.core.geotime.store.statistics.TimeRangeStatistic(type_name, field_name)
        super().__init__(java_ref, IntervalTransformer())
